contracts/periphery/src/execute.rs

```diff
@@ -1,125 +1,305 @@
-    sim_amount_desired.sort_by(|a, b| a.denom.cmp(&b.denom));
+    let deps_ref = deps.as_ref();
+    let sim = Simulator::new(&deps_ref, &pools, &swap_info, &core_portfolio.units);
+    let sim_res = sim
+        .estimate_index_for_input(
+            desired_asset.clone(),
+            Some(min_index_amount),
+            Some(min_index_amount),
+            None,
+        )?
+        .est
+        .unwrap();
 
-    let (swap_msgs, refund) = make_mint_swap_exact_out_msgs(
-        &deps.querier,
+    let swap_msgs = sim_res.sim_routes.to_msgs(
         &env.contract.address,
-        swap_info,
-        sim_amount_desired.clone(),
-        &max_input,
+        sim_res.max_token_in + Uint128::new(10000),
     )?;
 
     let finish_msg = WasmMsg::Execute {
         contract_addr: env.contract.address.to_string(),
-        msg: to_binary(&ExecuteMsg::FinishOperation {
+        msg: to_json_binary(&ExecuteMsg::FinishOperation {
             refund_to: info.sender.to_string(),
-            refund_asset: input_asset,
+            refund_asset: desired_denom,
         })?,
         funds: vec![],
     };
 
+    let mut mint_msg_funds = core_portfolio
+        .units
+        .into_iter()
+        .map(|(denom, unit)| coin((sim_res.est_min_token_out * unit).u128(), denom))
+        .collect::<Vec<_>>();
+    mint_msg_funds.sort_by(coin_sorter);
+
+    let mint_msg = core.call_with_funds(
+        core::ExecuteMsg::Mint {
+            amount: sim_res.est_min_token_out,
+            receiver: Some(info.sender.to_string()),
+            refund_to: Some(info.sender.to_string()),
+        },
+        mint_msg_funds,
+    )?;
+
+    let act_mint_amount = sim_res.est_min_token_out * deduct_fee(core_fee.mint_fee)?;
+    let act_mint_asset = coin(act_mint_amount.u128(), core_config.index_denom);
+
+    let resp = Response::new()
+        .add_messages(swap_msgs)
+        .add_message(mint_msg)
+        .add_message(finish_msg)
+        .add_attributes(vec![
+            attr("method", "mint_exact_amount_in"),
+            attr("executor", info.sender),
+            attr("input", desired_asset.to_string()),
+            attr("min_output", act_mint_asset.to_string()),
+        ]);
+
+    Ok(resp)
+}
+
+pub fn mint_exact_amount_out(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    core_addr: String,
+    index_amount: Uint128,
+    input_denom: String,
+    swap_info: Vec<SwapInfo>,
+) -> Result<Response, ContractError> {
+    // query to core contract
+    let core = IbcCore(deps.api.addr_validate(&core_addr)?);
+    let core_fee = core.get_fee(&deps.querier, None)?;
+    let core_config = core.get_config(&deps.querier, None)?;
+    let core_portfolio = core.get_portfolio(&deps.querier, None)?;
+
+    // input & output
+    let input_asset =
+        cw_utils::must_pay(&info, &input_denom).map(|v| coin(v.u128(), &input_denom))?;
+    let index_asset = coin(index_amount.u128(), &core_config.index_denom);
```

Comment:
ask; Is there no validation for the `input_denom` is an acceptable asset?

Reply:
Yep, there's no criteria to decide which denom is acceptable asset. So you can use periphery contract if you have feasible swap route for that.